src/TRoomDB.cpp

```diff
@@ -46,6 +46,35 @@ TRoomDB::TRoomDB(TMap* pMap)
     addArea(-1, mpMap->getDefaultAreaName());
 }
 
+TRoomDB::~TRoomDB()
+{
+    mBulkDeletionMode = true;
+    
+    // Get all pointers before clearing containers to prevent lookup issues
+    QList<TRoom*> const roomList = getRoomPtrList();
+    QList<TArea*> const areaList = getAreaPtrList();
+    
+    // Clear all containers first - this prevents individual destructors
+    // from trying to remove themselves from the containers (O(n²) behavior)
+    rooms.clear();
+    areas.clear();
+    entranceMap.clear();
+    areaNamesMap.clear();
+    hashToRoomID.clear();
+    roomIDToHash.clear();
+    
+    // Now delete all objects - their destructors will see mBulkDeletionMode=true
+    // and skip the expensive cleanup operations
+    for (auto room : roomList) {
+        delete room;
+    }
+    for (auto area : areaList) {
+        delete area;
+    }
+
+    mBulkDeletionMode = false;
```

Comment:
There is no point in resetting this flag - as it is about to be destroyed along-side all the other class members! :smile:
```suggestion
```